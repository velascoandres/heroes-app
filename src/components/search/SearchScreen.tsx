import React, { useMemo } from 'react';

import { parse } from 'querystring';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { IHero } from '../heroes/interfaces';
import { useHistory, useLocation } from 'react-router-dom';
import { getHeroeByName } from '../../selectors/getHeresByName';

export interface SearchFormState {
    searchText: string;
}

export interface SearchState {
    filteredHeroes: IHero[];
}

export const SearchScreen: React.FC = () => {
    const history = useHistory();

    const location = useLocation();
    const querystring = parse(location.search, '?');
    const { q } = Object.keys(querystring).length ? querystring : { q: '' }; // I put this because parse() returns => '{?q:'batman'}'

    const [formState, onChange] = useForm<SearchFormState>({ searchText: q as string });

    const { searchText } = formState;

    const filteredHeroes = useMemo(() => getHeroeByName(q as string), [q]);

    const handleFilter = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    };

    return (
        <div className="container">
            <div>
                <h1>Search Screen</h1>
            </div>

            <div className="row">
                <div className="col-md-5 col-sm-12">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={(e) => handleFilter(e)}>
                        <input
                            type="text"
                            id="searchText"
                            name="searchText"
                            value={searchText}
                            onChange={onChange}
                            placeholder="Find your hero"
                            className="form-control"
                        />

                        <button type="submit" className="btn m-1 btn-block btn-outline-primary">
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-md-7 col-sm-12">
                    <h4>Results...</h4>
                    <hr />
                    {q === '' && <div className="alert alert-info">Search a Hero</div>}

                    {q !== '' && !filteredHeroes.length && (
                        <div className="alert alert-warning">No results for {searchText}</div>
                    )}

                    {filteredHeroes.map((hero: IHero) => (
                        <HeroCard key={hero.id} hero={hero} />
                    ))}
                </div>
            </div>
        </div>
    );
};
