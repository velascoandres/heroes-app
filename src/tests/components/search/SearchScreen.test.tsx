import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { getMockRouterProps } from '../../test-utils/getMockRouter';
import { Router } from 'react-router-dom';

describe('Pruebas en <SearchScreen />', () => {
    test('debe de mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>,
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');
    });

    test('debe de mostrar a "batman" y el Input con el valor del "queryString"', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>,
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar un error si no se encuentra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123123']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>,
        );
        expect(wrapper.find('.alert-warning').text().trim()).toBe('No results for batman123123');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de llamar el push del history', () => {
        const historyMock = getMockRouterProps(null).history;

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Router history={historyMock}>
                   <Route path="/search" component={SearchScreen} />
                   <SearchScreen />
                </Router>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman',
            }
        });
        wrapper.find('form').simulate('submit');
        expect(historyMock.push).toHaveBeenCalledWith('?q=batman');
    });
});
