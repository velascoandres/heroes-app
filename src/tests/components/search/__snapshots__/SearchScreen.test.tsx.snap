// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de llamar el push del history 1`] = `
Array [
  "",
  <div
    className="container"
  >
    <div>
      <h1>
        Search Screen
      </h1>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-5 col-sm-12"
      >
        <h4>
          Search Form
        </h4>
        <hr />
        <form
          onSubmit={[Function]}
        >
          <input
            className="form-control"
            id="searchText"
            name="searchText"
            onChange={[Function]}
            placeholder="Find your hero"
            type="text"
            value=""
          />
          <button
            className="btn m-1 btn-block btn-outline-primary"
            type="submit"
          >
            Search...
          </button>
        </form>
      </div>
      <div
        className="col-md-7 col-sm-12"
      >
        <h4>
          Results...
        </h4>
        <hr />
        <div
          className="alert alert-info"
        >
          Search a Hero
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe de mostrar a "batman" y el Input con el valor del "queryString" 1`] = `
<div
  className="container"
>
  <div>
    <h1>
      Search Screen
    </h1>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-5 col-sm-12"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          id="searchText"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-md-7 col-sm-12"
    >
      <h4>
        Results...
      </h4>
      <hr />
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-7 col-sm-12"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-5 col-sm-12"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title w-100"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <hr />
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                className=""
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrar un error si no se encuentra el hero 1`] = `
<div
  className="container"
>
  <div>
    <h1>
      Search Screen
    </h1>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-5 col-sm-12"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          id="searchText"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-md-7 col-sm-12"
    >
      <h4>
        Results...
      </h4>
      <hr />
      <div
        className="alert alert-warning"
      >
        No results for 
        batman123123
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div
  className="container"
>
  <div>
    <h1>
      Search Screen
    </h1>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-5 col-sm-12"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          id="searchText"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-md-7 col-sm-12"
    >
      <h4>
        Results...
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
